core/modules/replica/tools/qserv-replica-master-http.cc

```diff
@@ -0,0 +1,380 @@
+            _workerEvictTimeoutSec
+        ).option(
+            "qserv-sync-timeout",
+            "the maximum number of seconds to wait before Qserv workers respond"
+            " to the synchronization requests before bailing out and proceeding"
+            " to the next step in the normal replication sequence. A value which"
+            " differs from 0 would override the corresponding parameter specified"
+            " in the Configuration.",
+            _qservSyncTimeoutSec
+        ).flag(
+            "qserv-sync-force",
+            "the flag which would force Qserv workers to update their list of replicas"
+            " even if some of the chunk replicas were still in use by on-going queries."
+            " This affect replicas to be deleted from the workers during the synchronization"
+            " stages",
+            _forceQservSync
+        ).option(
+            "replicas",
+            "the minimal number of replicas when running the replication phase"
+            " This number if provided and if it's not 0 will override the corresponding value found"
+            " in the Configuration.",
+            _numReplicas
+        ).option(
+            "iter",
+            "the number of iterations (a value of 0 means running indefinitely)",
+            _numIter
+        ).flag(
+            "purge",
+            "The binary flag which if provided would enable the optional purge algorithm in"
+            " the end of each replication cycle in order to eliminate excess replicas which"
+            " might get created by algorithm ran earlier in the cycle.",
+            _purge
+        ).flag(
+            "permanent-worker-delete",
+            "The flag would trigger the permanent removal of the evicted workers"
+            " from the configuration of the Replication system. Please, use"
+            " this option with caution as it may result in loosing some Configuration"
+            " data for the deleted workers",
+            _permanentDelete
+        );
+    }
+
+    /**
+     * @see Application::runImpl()
+     */
+    int runImpl() final {
+
+        LOGS(_log, LOG_LVL_INFO, "MASTER CONTROLLER  " << parser().serializeArguments());
+
+        _controller = Controller::create(serviceProvider());
+
+        // These threads should be running simultaneously
+
+        auto self = shared_from_base<MasterControllerHttpApp>();
+
+        _replicationThread = ReplicationThread::create(
+            _controller,
+            [self] (ControlThread::Ptr const& ptr) {
+                self->_failed = true;
+            },
+            _qservSyncTimeoutSec,
+            _replicationIntervalSec,
+            _numReplicas,
+            _numIter,
+            _purge
+        );
+        _replicationThread->start();
+
+        _healthMonitorThread = HealthMonitorThread::create(
+            _controller,
+            [self] (ControlThread::Ptr const& ptr) {
+                self->_failed = true;
```

Comment:
I went looking for the thread-safety-ness of `_failed`; I see that in use it only transitions from `false` to `true`, which would cause an abort sequence to start, so that's cool. From an interface & sequencing safety perspective, it might be cool if this was a var that could only change state in one direction. 
`class Failer {
public: 
    void doFail() { _failed = true; }
    bool isFailed() const { return _failed; } 
private:
    atomic bool _failed{false};
};

Reply:
Very interesting idea! I will definitively consider it.